functions/src/create-notification.ts

```diff
@@ -0,0 +1,172 @@
+  // TODO: Find replies to comments.
+  // TODO: Filter bets for only open bets
+  if (
+    sourceType === 'comment' ||
+    sourceType === 'answer' ||
+    sourceType === 'contract'
+  ) {
+    let reasonTextPretext = getReasonTextFromReason(sourceType, reason)
+
+    const notifyContractCreator = async (
+      userToReasonTexts: user_to_reason_texts
+    ) => {
+      if (shouldGetNotification(sourceContract.creatorId, userToReasonTexts))
+        userToReasonTexts[sourceContract.creatorId] = {
+          text: `${reasonTextPretext} your question`,
+          reason,
+        }
+    }
+
+    const notifyOtherAnswerersOnContract = async (
+      userToReasonTexts: user_to_reason_texts
+    ) => {
+      const answers = await getValues<Answer>(
+        firestore
+          .collection('contracts')
+          .doc(sourceContract.id)
+          .collection('answers')
+      )
+      const recipientUserIds = uniq(answers.map((answer) => answer.userId))
+      recipientUserIds.forEach((userId) => {
+        if (shouldGetNotification(userId, userToReasonTexts))
+          userToReasonTexts[userId] = {
+            text: `${reasonTextPretext} a question you submitted an answer to`,
+            reason,
+          }
+      })
+    }
+
+    const notifyOtherCommentersOnContract = async (
+      userToReasonTexts: user_to_reason_texts
+    ) => {
+      const comments = await getValues<Comment>(
+        firestore
+          .collection('contracts')
+          .doc(sourceContract.id)
+          .collection('comments')
+      )
+      const recipientUserIds = uniq(comments.map((comment) => comment.userId))
+      recipientUserIds.forEach((userId) => {
+        if (shouldGetNotification(userId, userToReasonTexts))
+          userToReasonTexts[userId] = {
+            text: `${reasonTextPretext} a question you commented on`,
+            reason,
+          }
+      })
+    }
+
+    const notifyOtherBettorsOnContract = async (
+      userToReasonTexts: user_to_reason_texts
+    ) => {
+      const betsSnap = await firestore
+        .collection(`contracts/${sourceContract.id}/bets`)
+        .get()
+      const bets = betsSnap.docs.map((doc) => doc.data() as Bet)
+      const recipientUserIds = uniq(bets.map((bet) => bet.userId))
+      recipientUserIds.forEach((userId) => {
+        if (shouldGetNotification(userId, userToReasonTexts))
+          userToReasonTexts[userId] = {
+            text: `${reasonTextPretext} a question you bet on`,
+            reason,
+          }
+      })
+    }
+
+    const getUsersToNotify = async () => {
+      const userToReasonTexts: user_to_reason_texts = {}
+      // The following functions modify the userToReasonTexts object in place.
+      await notifyContractCreator(userToReasonTexts)
+      await notifyOtherAnswerersOnContract(userToReasonTexts)
+      await notifyOtherCommentersOnContract(userToReasonTexts)
+      await notifyOtherBettorsOnContract(userToReasonTexts)
+      return userToReasonTexts
```

Comment:
I think it would be better to make all these functions immutable. Something like:
`const userToReasonTexts  = { ...await notifyContractCreator(userToReasonTexts), ...await notifyOtherAnswerersOnContract(userToReasonTexts), ... }`

Reply:
True, I knew that tho and am going to refactor them on the next iteration.